File: Sorts/Heap/heapSort.cpp
#include <iostream> 
#include <ctime>
#include <fstream>
const int MAXIMO = 10000;
using namespace std; 
  

void heapify(int arr[], int n, int i) 
{ 
    int largest = i; // Initialize largest as root 
    int l = 2 * i + 1; // left = 2*i + 1 
    int r = 2 * i + 2; // right = 2*i + 2 
  
 
    if (l < n && arr[l] > arr[largest]) 
        largest = l; 
 
    if (r < n && arr[r] > arr[largest]) 
        largest = r; 
  
    
    if (largest != i) { 
        swap(arr[i], arr[largest]); 
  
        
        heapify(arr, n, largest); 
    } 
} 
  

void heapSort(int arr[], int n) 
{ 
  
    for (int i = n / 2 - 1; i >= 0; i--) 
        heapify(arr, n, i); 
  
    
    for (int i = n - 1; i >= 0; i--) { 
        // Move current root to end 
        swap(arr[0], arr[i]); 
  
        heapify(arr, i, 0); 
    } 
} 

  
// Driver program 
int main() 
{ 
    int n;
    unsigned t0, t1;
 
    int cadena;
    ifstream file("entrada.txt");
    ofstream salida("heapSortCpp.txt");
    for (int i=50000;i<300001;i+=50000){
        int arreglo[ i ] = { 0 };
        for (int j=0;j<i;j++){
            file>>cadena;
            arreglo[j]=cadena;
                   
        }
        t0=clock();
        heapSort(arreglo,i);
        t1 = clock();

        double time = (double(t1-t0)/CLOCKS_PER_SEC);
        salida<<i<<"\t"<<time<<"\n";
        cout << "Execution Time: " << time << endl;
        
    }
    file.close();
    salida.close();

    cout << "Sorted array is \n";cin>>n; 
   
}

File: Sorts/Heap/heapSort.py
import numpy as np
import time 

def heapify(arr, n, i):
      largest = i
      l = 2 * i + 1
      r = 2 * i + 2
  
      if l < n and arr[i] < arr[l]:
          largest = l
  
      if r < n and arr[largest] < arr[r]:
          largest = r
  
      if largest != i:
          arr[i], arr[largest] = arr[largest], arr[i]
          heapify(arr, n, largest)
  
  
def heapSort(arr):
      n = len(arr)
  
      # max heap
      for i in range(n//2, -1, -1):
          heapify(arr, n, i)    
  
      for i in range(n-1, 0, -1):
          # Swap
          arr[i], arr[0] = arr[0], arr[i]
  
          # Heapify 
          heapify(arr, i, 0)
   
File = open("heapSortPython.txt","w")
entrada = open("entrada.txt",'r')
lineas = entrada.readlines()
x = [0]
for i in range(50000,300001,40000):
    print(i)
    for j in range(0,i):
        x.append(int(lineas[j]))
        
 
    # print ("array sin ordenar",x)
    time0=time.time()
    heapSort(x)
    timeF=time.time()
    tiempo=timeF-time0 
     #print ("array ordenado ",x)
     #print("tiempo de ordenacion =",tiempo)
    File.write(str(i)+"\t" + str(tiempo)+"\n")     
    x= [0]
File.close()
entrada.close()
